### PR TITLE
Feature/greedy pauli simp

### PR DESCRIPTION
**Pull request by yao-cqc**
# Description

Implement a greedy method for synthesising Pauli gadgets and Clifford tableaus
adapted from https://arxiv.org/abs/2103.08602 and https://arxiv.org/abs/2305.10966

# Related issues

Please mention any github issues addressed by this PR.

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented hard-to-understand parts of my code.
- [x] I have made corresponding changes to the public API documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have updated the changelog with any user-facing changes.

### CODE REVIEW

**Review by sjdilkes — CHANGES_REQUESTED** [review]
Looking excellent - I've only left a few comments, most regarding minor grammar changes.

File: tket/src/Transformations/GreedyPauliOptimisation.cpp [hunk 1]
```diff
@@ -0,0 +1,947 @@
...
+
+  // circuit used to iterate the original commands with flattened registers
+  Circuit circ_flat(circ);
+  circ_flat.flatten_registers();
+  std::vector<Command> commands = circ_flat.get_commands();
+  // extract the final clifford and the measurement circuits
+  for (const Command& cmd : commands) {
+    OpType optype = cmd.get_op_ptr()->get_type();
+    switch (optype) {
+      case OpType::Measure: {
```
[the remark this answers is not in the document]
Comment (yao-cqc) [reply tket/src/Transformations/GreedyPauliOptimisation.cpp:797]: This doesn't. I'm relying on ``synthesise_pauli_graph`` and ``NoMidMeasurePredicate`` at the pass level to detect mid-circuit measurments.

File: tket/src/Transformations/GreedyPauliOptimisation.cpp [hunk 2]
```diff
@@ -0,0 +1,947 @@
...
+      }
+    }
+  }
+  std::vector<std::vector<PauliExpNode>> rotation_sets;
+  std::vector<TableauRowNode> rows;
+  // use forward Tableau to update the paulis by commuting the tableau to the
+  // front
+  UnitaryTableau forward_tab = circuit_to_unitary_tableau(cliff);
+  // Tableau used for tracking Cliffords throughout the synthesis
+  // TODO: this can be potentially made redundant
```
[the remark this answers is not in the document]
Comment (yao-cqc) [reply tket/src/Transformations/GreedyPauliOptimisation.cpp:817]: This is just a reminder. I'm planning to do this in a future refactoring job.

**Review by sjdilkes — COMMENTED** [review]
LGTM

**Review by sjdilkes — APPROVED** [review]

**Pull request merged**